stairway/src/main/java/bio/terra/stairway/impl/MdcHelper.java

```diff
@@ -0,0 +1,76 @@
+package bio.terra.stairway.impl;
+
+import bio.terra.stairway.FlightContext;
+import java.util.Map;
+import java.util.Optional;
+import org.slf4j.MDC;
+
+/**
+ * Utility methods to make Stairway flight runnables context-aware, using mapped diagnostic context
+ * (MDC).
+ */
+class MdcHelper {
+
+  /** ID of the flight */
+  static final String FLIGHT_ID_KEY = "flightId";
+
+  /** Class of the flight */
+  static final String FLIGHT_CLASS_KEY = "flightClass";
+
+  /** Class of the flight step */
+  static final String FLIGHT_STEP_CLASS_KEY = "flightStepClass";
+
+  /** Direction of the step (START, DO, SWITCH, or UNDO) */
+  static final String FLIGHT_STEP_DIRECTION_KEY = "flightStepDirection";
+
+  /** The step's execution order */
+  static final String FLIGHT_STEP_NUMBER_KEY = "flightStepNumber";
+
+  /**
+   * @return the flightRunner modified to propagate the MDC from the calling thread and
+   *     flight-specific context to the child thread spawned to run the flight.
+   */
+  static Runnable withMdcAndFlightContext(Runnable flightRunner, FlightContext flightContext) {
+    // Save the calling thread's context
+    Map<String, String> contextMap = MDC.getCopyOfContextMap();
+    return () -> {
+      // Any leftover context on the thread will be fully overwritten:
+      MDC.setContextMap(Optional.ofNullable(contextMap).orElse(Map.of()));
+      // If the calling thread's context contains flight and step context from a parent flight, this
+      // will be overwritten below:
+      addFlightContextToMdc(flightContext);
+      removeStepContextFromMdc(flightContext);
```

Comment:
I don't understand removing step context here.

Reply:
When I added flight and step context to the MDC in WSM's MdcHook, I didn't account for subflights that were kicked off within a parent flight's step.  The initial logs emitted by subflights before the first step errantly included step-specific context from the calling flight, interspersed with flight-specific context from the new subflight.  https://github.com/DataBiosphere/terra-workspace-manager/pull/1636

What changes would help make this clear?

Reply:
Ah, I get it now.